File: AlgoPersonalStudy/src/backjoon/silver/BJ_11286.java
package backjoon.silver;

import java.util.*;
import java.io.*;

public class BJ_11286 {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());

        PriorityQueue<int[]> pq = new PriorityQueue<>(Comparator.comparingInt((int[] a) -> a[0]).thenComparingInt(a -> a[1]));

        StringBuilder sb = new StringBuilder();
        for (int i=0;i<n;i++){
            int x = Integer.parseInt(br.readLine());
            if (x == 0){
                if(!pq.isEmpty()) {
                    int[] polledData = pq.poll();
                    sb.append(polledData[1]).append("\n");
                }
                else{
                    sb.append(0).append("\n");
                }
            }
            else {
                pq.add(new int[]{Math.abs(x),x});
            }
        }
        System.out.println(sb.toString());
    }
}

File: AlgoPersonalStudy/src/_kotlin/backjoon/silver/BJ_11286.kt
package _kotlin.backjoon.silver

import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*
import kotlin.math.abs

class BJ_11286 {
    fun solution(){
        val br = BufferedReader(InputStreamReader(System.`in`))
        val n = Integer.parseInt(br.readLine())
        val pq = PriorityQueue(Comparator.comparingInt { a: IntArray -> a[0] }
            .thenComparingInt { a: IntArray ->
                a[1]
            })

        val sb = StringBuilder(br.readLine())
        for(i:Int in 0 until n){
            val x = Integer.parseInt(br.readLine())
            if(x == 0){
                if(!pq.isEmpty()){
                    val polledData =pq.poll()
                    sb.append(polledData[1]).append("\n")
                }
                else{
                    sb.append(0).append("\n")
                }
            }
            else{
                pq.add(intArrayOf(abs(x),x))
            }
        }
        println(sb.toString())
    }
}
fun main(){
    val solutionClass = BJ_11286()
    solutionClass.solution()
}